### PR TITLE
[NEW] [Apps-Engine] New Room events

### PR DESCRIPTION
**Pull request by d-gubert**
This will add new events that give apps the possibility of filtering users that will enter any room.

Closes https://github.com/RocketChat/Rocket.Chat/issues/17366

### CODE REVIEW

File: app/lib/server/functions/addUserToRoom.js [hunk 1]
```diff
@@ -27,6 +29,16 @@ export const addUserToRoom = function(rid, user, inviter, silenced) {
 		callbacks.run('beforeJoinRoom', user, room);
 	}
 
+	try {
```
Comment (rodrigok) [inline app/lib/server/functions/addUserToRoom.js:32]: I think this should run before the callbacks, WDYT?

Comment (d-gubert) [reply app/lib/server/functions/addUserToRoom.js:32]: My thinking to put it here was that "things in the core must run before the apps", but I don't really have a strong case about it. Why do you think it has to run before the callbacks?

Comment (rodrigok) [reply app/lib/server/functions/addUserToRoom.js:32]: Because the callbacks may execute things thinking that the room will be created

Comment (d-gubert) [reply app/lib/server/functions/addUserToRoom.js:32]: Solved [here](https://github.com/RocketChat/Rocket.Chat/pull/17656/files#diff-59771c3aa115644893a53436db956059)

**Pull request merged**